packages/metrics-discovery/src/index.ts

```diff
@@ -0,0 +1,27 @@
+import fs from 'fs'
+import { MetricsDiscovery } from './metrics-discovery'
+import { envVarsSchema } from './env-vars'
+import { sleep } from './utils'
+
+const PROMETHEUS_TARGETS_FILE = './prometheus/etc/config/targets.json'
+const SLEEP_DURATION_MS = 1000 * 60 * 5 // 5 minutes
+
+const run = async () => {
+    const envVars = envVarsSchema.parse(process.env)
+    const metricsDiscovery = MetricsDiscovery.init({
+        riverRpcURL: envVars.RIVER_RPC_URL,
+        env: envVars.ENV,
+    })
+
+    for (;;) {
+        console.info('Getting prometheus targets...')
+        const targets = await metricsDiscovery.getPrometheusTargets()
+        console.info('Writing prometheus targets...', targets)
+        fs.writeFileSync(PROMETHEUS_TARGETS_FILE, targets)
+        console.info(`Prometheus targets written to: ${PROMETHEUS_TARGETS_FILE}`)
+        console.info(`Sleeping for ${SLEEP_DURATION_MS} ms...`)
+        await sleep(SLEEP_DURATION_MS)
+    }
+}
+
+run().catch(console.error)
```

Comment:
Maybe better to use a watcher like fs.watch or chokadir:


import { promisify } from 'util';
import fs from 'fs';
import chokidar from 'chokidar';
import { MetricsDiscovery } from './metrics-discovery';
import { envVarsSchema } from './env-vars';

const writeFile = promisify(fs.writeFile);

const PROMETHEUS_TARGETS_FILE = './prometheus/etc/config/targets.json';

interface EnvVars {
    RIVER_RPC_URL: string;
    ENV: string;
}

const run = async () => {
    const envVars: EnvVars = envVarsSchema.parse(process.env);
    const metricsDiscovery = MetricsDiscovery.init({
        riverRpcURL: envVars.RIVER_RPC_URL,
        env: envVars.ENV,
    });

    const updateTargets = async () => {
        try {
            console.info('Getting prometheus targets...');
            const targets = await metricsDiscovery.getPrometheusTargets();
            console.info('Writing prometheus targets...', targets);
            await writeFile(PROMETHEUS_TARGETS_FILE, targets);
            console.info(`Prometheus targets written to: ${PROMETHEUS_TARGETS_FILE}`);
        } catch (error) {
            console.error('Error during metrics discovery or file write:', error);
        }
    };

    await updateTargets();

    const watcher = chokidar.watch(PROMETHEUS_TARGETS_FILE);

    watcher.on('change', async () => {
        console.info(`File change detected: ${PROMETHEUS_TARGETS_FILE}`);
        await updateTargets();
    });

    console.info('Watching for changes...');
};

run().catch(error => {
    console.error('Unhandled error in run function:', error);
    process.exit(1);
});

Reply:
@brianathere the metrics discovery script is write-only with respect to targets.json. it doesn't need to watch for anything. the only responsibility here is to fetch all nodes and write `targets.json`. prometheus then picks up the changes automatically from `targets.json` and updates its collectors.

good idea to make fs.write async though 👍